source/core/Array.cpp

```diff
@@ -1351,44 +1351,48 @@ struct ArrayReshapeStruct : public VarArgInstruction
 class ArrayIterator
 {
  private:
-    ArrayDimensionVector  _indexStack;
-    IntIndex _indexDim;
-    IntIndex*  dimensions;
-    IntIndex  _rank;
     TypedArrayCoreRef _array;
+    IntIndex  _rank;
+    IntIndex*  _dimensionLengths;
+    ArrayDimensionVector _indexes;
+
+ private:
+    void ResetIndexes() {
+        for (IntIndex i = 0; i < _rank; i++) {
+            _indexes[i] = 0;
+        }
+    }
+
  public:
     explicit ArrayIterator(TypedArrayCoreRef array) {
         _array = array;
         _rank = array->Rank();
-        _indexDim = 0;
-        dimensions = array->DimensionLengths();
-        for (IntIndex i = 0; i < _rank; i++) {
-            _indexStack[i] = 0;
-        }
+        _dimensionLengths = array->DimensionLengths();
+        ResetIndexes();
     }
+
     void* Begin() {
-        for (IntIndex i = 0; i < _rank; i++) {
-            _indexStack[i] = 0;
-        }
-        _indexDim = 0;
-        return _array->BeginAtND(_rank, _indexStack);
+        ResetIndexes();
+        return _array->BeginAtND(_rank, _indexes);
     }
+
     void* Next() {
-        _indexStack[_indexDim]++;
-        if (dimensions[_indexDim] <= _indexStack[_indexDim]) {
-            while (dimensions[_indexDim] <= _indexStack[_indexDim]) {
-                _indexDim++;
-                _indexStack[_indexDim]++;
-            }
-            if (_indexDim >= _rank) {
-                return NULL;
-            }
-            for (IntIndex i = 0; i < _indexDim; i++) {
-                _indexStack[i] = 0;
-                _indexDim = 0;
+        IntIndex dimensionIndex = 0;
+        while (dimensionIndex < _rank) {
+            _indexes[dimensionIndex]++;
+            if (_indexes[dimensionIndex] >= _dimensionLengths[dimensionIndex]) {
```

Comment:
I disagree.  The expression to check if an index is in bounds should be a form of the invariant "index < length".   This expression is the straight-forward inverse of that to check if it's out of bounds and it seems more rigorous and readable to me.   The two are equivalent in the context of this loop, but in general, why would you prefer a weaker test?